pkg/services/object/put/validation.go

```diff
@@ -20,9 +21,34 @@ type validatingTarget struct {
 	hash hash.Hash
 
 	checksum []byte
+
+	maxPayloadSz uint64 // network config
+
+	payloadSz uint64 // payload size of the streaming object from header
+
+	writtenPayload uint64 // number of already written payload bytes
 }
 
+// errors related to invalid payload size
+var (
+	errExceedingMaxSize = errors.New("payload size is greater than the limit ")
+	errWrongPayloadSize = errors.New("wrong payload size")
+)
+
 func (t *validatingTarget) WriteHeader(obj *object.RawObject) error {
+	t.payloadSz = obj.PayloadSize()
+	chunkLn := uint64(len(obj.Payload()))
+
+	// check chunk size
+	if chunkLn > t.payloadSz {
+		return errWrongPayloadSize
+	}
+
+	// check payload size limit
+	if t.payloadSz > t.maxPayloadSz {
```

Comment:
Wouldn't it better to do this?
```suggestion
	if chunkLn > t.maxPayloadSz {
```

As far as I understand, client can send 2GiB of payload in stream with correct object size in header and this check will fail, or am I missing something?

Reply:
`chunkLn` should not be GE `payloadSz` by condition above. This condition checks that the client does not start streaming a obviously larger object. Your condition is satisfied automatically by the transitivity of the `>` operator.

Reply:
`validatingTarget.WriteHeader` is not executed against unfinished objects, I missed that. Therefore my comment is invalid.